### PR TITLE
NIFI-6171 always send email scope for OIDC

### PR DESCRIPTION
**Pull request by SimonLinder**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [X] Is your initial contribution a single, squashed commit?

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [X] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by mcgilman** [comment]
Thanks for filing this JIRA and creating this PR @SimonLinder! I also appreciate the write up in the JIRA. It makes this much easier for me to review.

- Good catch on the incorrect null check.
- If we can identify the fact that the IdP cannot support what we require, it's better to throw the exception to prevent startup. This is why there are so many checks in the constructor.
- I didn't realize the claims specified in the call to the Authorization endpoint had any impact on a subsequent call to details returned from a UserInfo endpoint. 
- From my initial testing with various IdP implementations, not all supported the email claim in their Authorization endpoint. Many did, however, support returning it in the subsequent call to the UserInfo endpoint.
- There were others though, that did not support an email field in their UserInfo endpoint. We have received requests [1] to allow for additional configuration over what field we use to identify a user.

I am concerned with requiring the email scope be part of the Authorization request would be too limiting. From your experience, would you have any guidance on how we could try to remain as flexible, identify users using any field (from the UserInfo endpoint), and work against as many IdP's as possible?

[1] https://github.com/apache/nifi/pull/2346

**Comment by SimonLinder** [comment]
Hi @mcgilman 
thanks for your comment on my issue.

You are right about the exceptions, thrown in the constructor. But in this particular case, as stated in the spec of the discovery end, it is not mandatory to send the supported scopes, thus if the scope **email** is not returned, it is not correct to suppose that this scope is not supported. Therefore I would not make that check within the constructor. Same goes with the **openid** scope.
I came across this issue with an implementation of the [IdentityServer](https://identityserver.io/), where all **supported_scopes** where set to null, which is totally fine, but not RECOMMENDED.

I can follow your concern about removing the **lookupEmail()** function. And I think you're right about this. Therefore I would suggest to leave that function as is, be it as a fallback, but removing the check in the constructor (as I already did) and always adding the **email** scope (as I already did).

If you're agreeing to my suggestion I would change that accordingly and adjust my commit.

Cheers
Simon

**Comment by mcgilman** [comment]
@SimonLinder I think I'd need to try it out. I recall some IdP's rejecting requests when the scope contained `email` when it was explicitly not supported. I think the only scope that is always supported is 'openid'. Hence all the conditional inclusion of the `email` scope in the current codebase.

At this point, I am worried about regressions in behavior due to this change. I understand why the current codebase fails when the `supported_scopes` is not specified. With that fix in place, did things start working against your IdP?

**Comment by SimonLinder** [comment]
@mcgilman Yes, with this fix in place, I could work against my IDP.
I further investigated the issue.
So first of all, [here](https://auth0.com/docs/scopes/current/oidc-scopes) you can find the explanation about my observation concerning the call to the userinfo endpoint:
Once the user authorizes the requested scopes, the claims are returned in an ID Token and are also available through the /userinfo endpoint.
So the **email** scope is only available via the userinfo endpoint if it was requested previously. So this would confirm my suggestion to remove the lookEmail() function; beside the work-around for #NIFI-4698.

You're right about the required scope. Only the **openid** scope is required to be sent when requesting the token. But NiFi is highly dependent on the existence of the email value. Otherwise NiFi could not verify the user that wants to log in.
So what would happen if the user requests the **email** scope and the IDP won't support it? If the request really throws an exception or simply doesn't return the email value? An exception is caught by the try-catch block, an empty email value would lead to a request to the userinfo endpoint, that probably will also have no result, thus the user cannot login to NiFi (using this IDP). So the selected IDP might be not the correct choice.

**Comment by SimonLinder** [comment]
@mcgilman Sorry, did some wired GIT stuff, because I accidentally pushed stuff to a branch "nifi-6171", but I wanted to use "NIFI-6171". That's why there are wrong commits. Current state of branch NIFI-6171 is my initial commit, including all changes that were made in upstream/master.

**Comment by mcgilman** [comment]
@SimonLinder Please rebase your branch and remove the merge commits. Happy to evaluate it against the IdP's I've used in the past.

**Comment by SimonLinder** [comment]
@mcgilman Done. I'm looking forward to your evaluation.

**Comment by mcgilman** [comment]
@SimonLinder Thanks for updating your branch. I just tried the PR against the mitreid-connect IdP [1] and it is exhibiting the same behavior that I recall seeing during the initial development of this capability. The token request that we perform to get the id and access token return successfully. However, despite the presence of the `email` scope the id token does not necessarily contain the claims that correspond to the authorized scopes. It does, however, as you pointed out, make those claims available via subsequent calls to the UserInfo endpoint. 

Looking at the spec for the id token [2] it appears that the id token may contain additional claims. I think that is what we're seeing with the Google Idp and possibly the implementation you're using. I think we need to retain the `lookupEmail` method to restore the capabilities for implementations that do not include additional claims in the token response.

I propose the following:

1) Always include both `openid` and `email` scopes.
2) Remove unnecessary checks for scopes in the constructor.
3) Retain `lookupEmail` for invoking the UserInfo endpoint.

I'd also like to get your opinion on optionally introducing the following:

4) Add a property to `nifi.properties` to allow the nifi admin to configure additional scopes (e.g. `profile`). These would be included in addition to `openid` and `email`.
5) Add a property to `nifi.properties` to allow the nifi admin to configure the claim that we should use to identify the user. This would allow folks to configure any claim to identify the user and not be dependent on `email`. This property would default to `email` to retain backward compatibility. However, the nifi admin could ultimately choose any claim as long as it was available through the configured scopes.

These changes should be pretty straight forward and I am happy to review if you're interested in contributing them too. If not, I can file follow on JIRAs for these improvements. 

[1] https://github.com/mitreid-connect/OpenID-Connect-Java-Spring-Server
[2] https://openid.net/specs/openid-connect-core-1_0.html#IDToken

**Comment by SimonLinder** [comment]
@mcgilman Thanks for testing the code against other IDPs.
I agree in your proposal and will change that.
I also will make the changes to the `nifi.properties` as you suggested. I think these features are very helpful for making OIDC handling in NiFi more flexible.
I'll come back to you when I'm done with the implementation.

**Comment by mcgilman** [comment]
@SimonLinder Sounds great. Thanks again for contributing!

**Comment by SimonLinder** [comment]
@mcgilman I'm done with my changes. Please review...

**Comment by mcgilman** [comment]
@SimonLinder Sorry I have been traveling the past week. Will try to check out your latest updates today. Thanks!

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/src/main/java/org/apache/nifi/web/security/oidc/StandardOidcIdentityProvider.java [hunk 1]
```diff
@@ -307,17 +296,22 @@ public String exchangeAuthorizationCode(final AuthorizationGrant authorizationGr
...
+
                     // extract the bearer access token
                     final BearerAccessToken bearerAccessToken = oidcTokens.getBearerAccessToken();
                     if (bearerAccessToken == null) {
                         throw new IllegalStateException("No access token found in the ID tokens");
                     }
 
                     // invoke the UserInfo endpoint
-                    email = lookupEmail(bearerAccessToken);
+                    identity = lookupEmail(bearerAccessToken);
```
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/src/main/java/org/apache/nifi/web/security/oidc/StandardOidcIdentityProvider.java]: Sorry if the earlier comments weren't clear. I thought we should retain `lookupEmail` in order to invoke the UserInfo endpoint. What I did a poor job relaying that I think it should be updated to look up the user identity as configured in `nifi.properties`. So if I configured an additional scope of `profile` and a claim of `preferred_username` we should be extracting that claim fro the UserInfo response. Assuming we make these changes, it probably makes sense to just rename this method `lookupIdentity`.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/src/main/java/org/apache/nifi/web/security/oidc/StandardOidcIdentityProvider.java [hunk 2]
```diff
@@ -307,17 +296,22 @@ public String exchangeAuthorizationCode(final AuthorizationGrant authorizationGr
...
-                // attempt to extract the email from the id token if possible
-                String email = claimsSet.getStringClaim(EMAIL_CLAIM_NAME);
-                if (StringUtils.isBlank(email)) {
+                // attempt to extract the configured claim to access the user's identity; default is 'email'
+                String identity = claimsSet.getStringClaim(properties.getOidcClaimIdentifyingUser());
+                if (StringUtils.isBlank(identity)) {
+                    // explicitly try to get the identity from the UserInfo endpoint with the 'email' claim
+                    logger.warn("The identity of the user was tried to get with the claim '" +
+                            properties.getOidcClaimIdentifyingUser() + "'. The according additional scope is not " +
+                            "configured correctly. Trying to get it with the 'email' claim.");
```
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/src/main/java/org/apache/nifi/web/security/oidc/StandardOidcIdentityProvider.java]: Assuming we update `lookupEmail` we should probably adjust this log message.

**Comment by SimonLinder** [comment]
@mcgilman I changed the lookupEmail() function so that it now tries to get whatever claim was configured in nifi.properties. I also think that this will make [NIFI-4698](https://github.com/apache/nifi/pull/2346) obsolete.

**Comment by mcgilman** [comment]
Thanks @SimonLinder! This has been merged to master. Changes look great. Thanks for working through this with me.

**Pull request closed without merging**